packages/ext-taquito/src/tzktReadProvider.ts

```diff
@@ -0,0 +1,235 @@
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+//@ts-nocheck
+import 'cross-fetch/polyfill';
+import {
+  BlockResponse,
+  EntrypointsResponse,
+  MichelsonV1Expression,
+  SaplingDiffResponse, ScriptedContracts,
+} from '@taquito/rpc';
+import {
+  BigMapQuery,
+  BlockIdentifier,
+  SaplingStateQuery,
+  TzReadProvider
+} from '@taquito/taquito';
+import {
+  accountsGetBalance,
+  accountsGetBalanceAtLevel,
+  accountsGetByAddress,
+  accountsGetCounter, bigMapsGetKey, Block, blocksGet,
+  blocksGetByHash,
+  blocksGetByLevel,
+  blocksGetCount,
+  contractsGetCode, contractsGetEntrypoints,
+  contractsGetRawStorage, headGet, Protocol, protocolsGetByCycle,
+  protocolsGetCurrent
+} from "@tzkt/sdk-api";
+import {BigNumber} from 'bignumber.js';
+
+
+
+export class TzktReadProvider implements TzReadProvider {
+  constructor(private readProvider: TzReadProvider) {}
+
+  private readonly liveBlocksLimit = 120;
+
+  private async _getBlockLevel(block: BlockIdentifier): Promise<number> {
+    if (typeof block === 'number') {
+      return block
+    }
+
+    if (this._blockIdIsHead(block)) {
+      return blocksGetCount();
+    }
+
+    if (String(block).includes('head~')) {
+      const n = Number(String(block).split('head~')[1])
+      const count = await blocksGetCount();
+
+      return Number(count) - n
+    }
+
+    const {level} = await blocksGetByHash(block);
+    return level || 0
+  }
+
+  private _blockIdIsHash(block: BlockIdentifier) {
+    return String(block)[0] === 'B'
+  }
+
+  private _blockIdIsHead(block: BlockIdentifier) {
+    return !block || block === 'head'
+  }
+
+  async getEntrypoints(contract: string): Promise<EntrypointsResponse> {
+    const response = await contractsGetEntrypoints(contract, {json: false, micheline: true});
+
+    const entypoints = response.map((entypoint) => {
+      return [entypoint.name, entypoint.michelineParameters]
+    });
+
+    return  {
+      entrypoints: Object.fromEntries(entypoints)
+    }
+  }
+
+  async getScript(address: string, block: BlockIdentifier): Promise<ScriptedContracts> {
+    if (this._blockIdIsHead(block)) {
+      return {code: await contractsGetCode(address), storage: this.getStorage(address, block)};
+    }
+    const blockLevel = await this._getBlockLevel(block);
+
+    return {code: await contractsGetCode(address, {level: blockLevel}), storage: this.getStorage(address, block)};
```

Comment:
Possible double call to "_getBlockLevel" inside this.getStorage

hint: extract this._getStorageAtLevel(address, level)